File: crates/libs/lib-core/src/models/planet.rs
use mongodb::bson::oid::ObjectId;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Deserialize, Serialize)]
pub enum PlanetThreat {
    #[serde(rename = "radiation")]
    Radiation,
    #[serde(rename = "toxins")]
    Toxins,
    #[serde(rename = "freezing")]
    Freezing,
    #[serde(rename = "heat")]
    Heat,
}

impl ToString for PlanetThreat {
    fn to_string(&self) -> String {
        match self {
            PlanetThreat::Radiation => "radiation",
            PlanetThreat::Toxins => "toxins",
            PlanetThreat::Freezing => "freezing",
            PlanetThreat::Heat => "heat",
        }
        .to_string()
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Planet {
    pub _id: ObjectId,
    pub system_id: ObjectId,
    pub threat: PlanetThreat,
}

File: api/models/planet.go
package models

import (
	"database/sql/driver"

	"github.com/google/uuid"
)

type PlanetThreat string

const (
	RADIATION PlanetThreat = "RADIATION"
	TOXINS    PlanetThreat = "TOXINS"
	FREEZING  PlanetThreat = "FREEZING"
	HEAT      PlanetThreat = "HEAT"
)

func (ct *PlanetThreat) Scan(value interface{}) error {
	*ct = PlanetThreat(value.([]byte))
	return nil
}

func (ct PlanetThreat) Value() (driver.Value, error) {
	return string(ct), nil
}

type Planet struct {
	ID       uuid.UUID `gorm:"type:uuid;default:gen_random_uuid()"`
	SystemID uuid.UUID
	System   System
	Threat   PlanetThreat `gorm:"type:planet_threat"`
}
